liblsp/OutputGenerator.cpp

```diff
@@ -0,0 +1,208 @@
+#include <liblsp/OutputGenerator.h>
+#include <libsolutil/Visitor.h>
+#include <variant>
+
+namespace lsp {
+
+using namespace std;
+
+OutputGenerator::NotificationInfo OutputGenerator::operator()(protocol::Notification const& _message)
+{
+	return visit(*this, _message);
+}
+
+OutputGenerator::NotificationInfo OutputGenerator::operator()(protocol::CancelRequest const& _message)
+{
+	Json::Value params;
+
+	visit(
+		solidity::util::GenericVisitor{
+			[&](int _id) {
+				params["id"] = _id;
+			},
+			[&](string const& _id) {
+				params["id"] = _id;
+			}
+		},
+		_message.id
+	);
+
+	return {"$/cancelRequest", params};
+}
+
+Json::Value OutputGenerator::toJson(Range const& _range)
+{
+	Json::Value json;
+	json["start"]["line"] = _range.start.line;
+	json["start"]["character"] = _range.start.column;
+	json["end"]["line"] = _range.end.line;
+	json["end"]["character"] = _range.end.column;
+	return json;
+}
+
+OutputGenerator::NotificationInfo OutputGenerator::operator()(protocol::PublishDiagnosticsParams const& _params)
+{
+	Json::Value params;
+
+	params["uri"] = _params.uri;
+
+	if (_params.version)
+		params["version"] = _params.version.value();
+
+	params["diagnostics"] = Json::arrayValue;
+	for (protocol::Diagnostic const& diag: _params.diagnostics)
+	{
+		Json::Value jsonDiag;
+
+		jsonDiag["range"] = toJson(diag.range);
+
+		if (diag.severity.has_value())
+			jsonDiag["severity"] = static_cast<int>(diag.severity.value());
```

Comment:
Why do you need the `static_cast` here and below?

Reply:
ein class to int. Or can JSON-CPP deal with this? I do not know m i can try.